components/log-viewer-webui/client/src/api/query.js

```diff
@@ -0,0 +1,55 @@
+import axios, {AxiosError} from "axios";
+
+
+/**
+ * @typedef {number} QueryLoadState
+ */
+let enumQueryLoadState;
+/**
+ * Enum of query loading state.
+ *
+ * @enum {QueryLoadState}
+ */
+const QUERY_LOAD_STATE = Object.freeze({
+    SUBMITTING: (enumQueryLoadState = 0),
+    WAITING: ++enumQueryLoadState,
+    LOADING: ++enumQueryLoadState,
+});
+
+/**
+ * Submits a job to extract a segment of the original file, which contains a given log event index,
+ * into CLP's IR format for viewing in the Log Viewer.
+ *
+ * @param {number|string} origFileId The ID of the original file to extract IR from
+ * @param {number} logEventIx The index of the log event
+ * @param {Function} onQueryStateChange Callback to set query state.
+ * @param {Function} onErrorMsg Callback to set error message.
+ */
+const submitExtractIrJob = async (origFileId, logEventIx, onQueryStateChange, onErrorMsg) => {
+    try {
+        const {data} = await axios.post("/query/extract-ir", {
+            msg_ix: logEventIx,
+            orig_file_id: origFileId,
+        });
+
+        onQueryStateChange(QUERY_LOAD_STATE.LOADING);
+
+        const innerLogEventIx = logEventIx - data.begin_msg_ix + 1;
+        window.location = `/log-viewer/index.html?filePath=/ir/${data.path}` +
+            `#logEventIdx=${innerLogEventIx}`;
+    } catch (e) {
+        let errorMsg = "Unknown error.";
+        if (e instanceof AxiosError) {
+            errorMsg = e.message;
+            if ("undefined" !== typeof e.response) {
+                errorMsg = e.response.data.message ?? e.response.statusText;
+            }
```

Comment:
> really we're implementing an if-else

Right, depending on whether the request was sent successfully and whether the server returns a valid response, we pick the error message (from one of the 3 places) that best represents the error. 

Let's continue the discussion with these as references:
> 1. If the caught error is caused by Axios, the error is of type `AxiosError` in all cases.
> 2. `message` , `name` , and `code` seem always defined even when the request doesn't reach the server.
      1. `message` is a high-level description of what happened
      2. `name` is name of the extended `Error` class, usually `AxiosError` .
      3. `code` is specific to the error. When the system call fails it could be a errno code; when it is a request error it’s a code provided by Axios. Good for debugging?
> 3. `config.url` is good information
> 4. When `reponse` is available:
      1. `response.status` contains the HTTP status code: 404, …
      2. `response.statusText` contains the HTTP status text: 404
      3. `response.data` contains the HTTP response body

> Can we rewrite it as an if-else

Sure

Reply:
If we're obsessed enough with syntactic sugar, we can write it as
```
    errorMsg = e.response?.data?.message ?? e.response?.statusText ?? e.message;
```
(only if it doesn't give us more "?" when reading the code; pun intended lol

Reply:
😛 I think in this case, fewer "?" is better, lol.